### PR TITLE
KEYCLOAK-13639. Added metrics and custom healthcheck endpoints.

### PR DESCRIPTION
**Pull request by miquelsi**
This PR adds the following endpoints:

- http://server:port/health
- http://server:port/health/live 
- http://server:port/health/ready
- http://server:port/auth/metrics 

The /ready endpoint has a custom database healthcheck using Agroal metrics.

This functionality is disabled by default. We can enable it via configuration using:

`java -jar server/target/lib/quarkus-run.jar config --metrics-enabled=true`

### CODE REVIEW

**Review by slaskawi — DISMISSED** [review]

**Comment by slaskawi** [comment]
@pedroigor May I ask you to review this? This is the Healthcheck Endpoint I told you about. We will be building more on top of this (see the backlog of the Cheetah Team).

**Review by slaskawi — CHANGES_REQUESTED** [review]
I checked the code and I believe we need some changes:
- The implementation as well dependencies should go to `quarkus/runtime`. This module contains most of the existing Quarkus dependencies as well as the runtime code.
- The BOM should be removed. It's already included in the main `pom.xml`.
- (in the future) The test should also probably go to `quarkus/runtime` module. The other alternative is our base testsuite.

**Comment by pedroigor** [comment]
@slaskawi I share all your points. The `server` module should be as simple as possible because in the future people should be able to create their own custom  Keycloak server based on our extension (runtime + deployment modules).

I've also replied to the thread about the runtime issues when accessing the health endpoint.

**Comment by miquelsi** [comment]
Updated with the suggested changes. Thanks @slaskawi and @pedroigor!

**Review by pedroigor — DISMISSED** [review]

File: services/src/main/java/org/keycloak/transaction/JtaTransactionWrapper.java
Comment (slaskawi) [inline services/src/main/java/org/keycloak/transaction/JtaTransactionWrapper.java:91]: Nice work!

**Review by slaskawi — DISMISSED** [review]
LGTM!

**Comment by slaskawi** [comment]
@miquelsi I've noticed this PR is still a Draft. We have two choices at this point:
- Convert it to a "real" PR and ask @pedroigor to merge it. We have all the approvals.
- Add more features and keep it as a Draft for now. We can merge it once all the bits fall in place. 

I definitely would like to see some tests as a next step. Testing failure scenarios might be very tricky. We will need to instruct Agroal to "disconnect" somehow from the database.

**Comment by pedroigor** [comment]
@slaskawi For me we can just merge it. We have a lot of room for preview work in Quarkus and as long as it does not break anything, it should be fine.

I would just ask @mposolda review that change to the transaction bits to make sure it is OK as it also related to code run by the main dist. I think it is fine, but let's see what he thinks.

**Comment by mposolda** [comment]
@pedroigor @slaskawi Transaction change looks ok to me.

**Comment by miquelsi** [comment]
I've been trying to add a simple Quarkus test but without success, so I think we can go ahead and merge this as a first step and keep on working. @pedroigor can you proceed?

**Comment by slaskawi** [comment]
Let's merge this one @pedroigor and @miquelsi .

I have prepared another PR with the improved Health check (passive + active mode): https://github.com/keycloak/keycloak/pull/7557

**Comment by pedroigor** [comment]
@miquelsi Could you rebase ? I'll merge it today.

**Comment by miquelsi** [comment]
> @miquelsi Could you rebase, please ? I'll merge it today.

Rebased, you can go ahead. Thanks!

**Comment by pedroigor** [comment]
@miquelsi After talking with @slaskawi about how we are exposing health endpoint, I'll hold it a bit more. Sorry.

We should first figure out how we are going to help people to not expose this endpoint for the outside world, but only for other applications/services running within the internal network.

It might as simple as relying on a proxy/ingress to restrict access to that endpoint. Not sure, more discussions need to happen.

**Comment by slaskawi** [comment]
Yes, this relates to [KEYCLOAK-15887](https://issues.redhat.com/browse/KEYCLOAK-15887), which is currently investigated by @ASzc. From my side - I'll look into Ingress and Route support for blacklisting users from being exposed to the outside world.

**Comment by pedroigor** [comment]
@slaskawi @miquelsi I've some changes to enable/disable health and metrics. There is no way to do that through their respective extensions so we would need something in our filter.

Basically, the idea is to use two properties like that:

```
metrics.enabled=true
health.enabled=true
```

To enable/disable these capabilities.

But there are some caveats that I'm still looking such as if this configuration should force a re-configuration of the server due to possible optimizations when blocking requests to those endpoints or not.

**Comment by pedroigor** [comment]
@slaskawi @miquelsi Here is the branch with some changes to include metrics and health support. Based on this change set.

* Added a `metrics.enabled` property to enable/disable metrics
* Added a `health.enabled` property to enabled/disable health

These two properties can only be set when configuring the server as follows:

```
kc.sh config --metrics-enabled=true --health-enabled=true
```

And then run the server:

```
kc.sh
```

By default they are disabled and as such both endpoints will respond with a `404` status code.

The reason why we need the configuration step is due to the fact that both health and metrics extensions require relies on build time configuration to setup things.

For us, this also means a bit of optimization at runtime to fail-fast when these capabilities are disabled. Ideally we should be able to remove all endpoints and dependencies from the resulting server image after the configuration step above. But that is not possible due to how those extensions work.

Another thing is related to how `KeycloakReadyHealthCheck` is implemented. In order to work properly you always need both capabilities enabled, otherwise if you don't enable `metrics`, Agroal won't configure datasources to expose their metrics thus making our health check to give a wrong status (`DOWN`). We need to discuss more this one.

**Comment by slaskawi** [comment]
> Another thing is related to how KeycloakReadyHealthCheck is implemented. In order to work properly, you always need both capabilities enabled, otherwise if you don't enable metrics, Agroal won't configure datasources to expose their metrics thus making our health check to give a wrong status (DOWN). We need to discuss more this one.

@pedroigor This has been done intentionally this way. Perhaps whenever we turn the healthcheck on, we should always turn the metrics on as well? Could you please give us a hint how to achieve this behavior?

**Comment by miquelsi** [comment]
> @pedroigor This has been done intentionally this way. Perhaps whenever we turn the healthcheck on, we should always turn the metrics on as well? Could you please give us a hint how to achieve this behavior?

Perhaps using only one property instead of two?

**Comment by pedroigor** [comment]
@slaskawi @miquelsi Yeah, I think a single option should be better. But I wasn't sure whether or not there cases where you don't want to expose metrics, for instance.

Based on my branch, that should be quite easy to do. We just need to figure out a good name.

In a nutshell, this is the part of the [code](https://github.com/pedroigor/keycloak/commit/f1effe0cd3e55931786cacc52770bbd85187ef54#diff-28080a2bdd954e526742b0efd4bc93ea2384625554ee23349ca1d0637357555fR155) that deals with mapping our properties to Quarkus.

So we basically set the value of `metrics.enabled` to `quarkus.datasource.metrics.enabled`. If we need to enable metrics for another extension, then we would do the same.

**Comment by pedroigor** [comment]
@slaskawi @miquelsi So, how do you plan to proceed? Do my changes make sense for you?

**Comment by slaskawi** [comment]
@pedroigor @miquelsi @stianst I asked the Quarkus and SmallRye Health maintainers a few questions offline (you all should be on cc). It seems there's a growing demand for enhancing the security aspect around metrics and health endpoints, see https://github.com/quarkusio/quarkus/issues/7893

So to me, it's very likely that once the Quarkus issue is implemented, we could enable both metrics and health endpoints by default. So we're talking about a temporary short-to-mid term solution here.

Having said that, I'm OK with going the easiest path and forcing users to specify both `--metrics-enabled=true --health-enabled=true`. The only enhancement I can see is that we could somehow try to detect whether metrics are turned on and then perform the passive check. Otherwise we should always perform the active one. Perhaps, we're lucky and it already works this way. If `agroalDataSource.getMetrics().activeCount()` returns `0` or `-1` if metrics are disabled, we don't need any additional changes. The code will work fine. @miquelsi May I kindly ask you to check this case?

**Comment by pedroigor** [comment]
Yeah, production-wise both capabilities should always be enabled as you do need to monitor your env and use health for checks.

One of the main ideas around Keycloak on Quarkus is to make configuration easier while providing good defaults. So I can see we defining a `prod` profile to enable metrics plus anything else (including security policies for clients, etc) that should be enforced when running in production.

**Comment by miquelsi** [comment]
Hi @pedroigor, @slaskawi 

I've integrated the changes you both suggested. Since there is still an ongoing discussion about how to secure health and metrics endpoints, that is still not covered here.

I have kept only one config parameter `metrics.enabled`. It affects both `/metrics` and `/health` endpoints, so perhaps we can use a better naming. What do you suggest? In any case the behavior is documented. For enabling them we will use

`java -jar server/target/lib/quarkus-run.jar config --metrics-enabled=true`

and default option is false. Please take a look and add any suggestions you may have.

**Review by pedroigor — COMMENTED** [review]
LGTM. I still have some minor improvements that may help to improve how we install/enable metrics such as:

* Use a build item to install our health check
* Investigate more if there is a better way to enable/disable metrics/health by forcing their respective extension to install anything

Other than that, it is a good start. So, let me know how you want to proceed.

**Comment by pedroigor** [comment]
There are some unexpected errors in CI.

**Review by slaskawi — APPROVED** [review]
LGTM, but @pedroigor needs to approve this. Perhaps you two could sync offline and discuss the changes needed here. We'd like to get this one in soon.

**Comment by miquelsi** [comment]
> LGTM. I still have some minor improvements that may help to improve how we install/enable metrics such as:
> 
> * Use a build item to install our health check
> * Investigate more if there is a better way to enable/disable metrics/health by forcing their respective extension to install anything
> 
> Other than that, it is a good start. So, let me know how you want to proceed.

I would go ahead merging this and include further improvements in a next stage. Can you please review the CI failures? I think it might not be related to the changes here.

**Comment by pedroigor** [comment]
@miquelsi Yeah, it does not seem related. Running again. But yeah, LGTM.

**Review by pedroigor — APPROVED** [review]

**Comment by pedroigor** [comment]
@miquelsi Could you rebase, please? Quarkus tests should not be required and I'm unable to merge.

**Review by mposolda — APPROVED** [review]

**Pull request merged**
